src/arch/util/machine-smp.C

```diff
@@ -589,21 +531,60 @@ static void CmiDestroyLocks(void)
 
 #if !CMK_SHARED_VARS_UNAVAILABLE
 
+class Barrier {
+public:
+  Barrier(const Barrier&) = delete;
+  Barrier& operator=(const Barrier&) = delete;
+
+  explicit Barrier(unsigned int count) :
+    curCount(count), barrierCount(count), curGeneration(0) {
+  }
+
+  void wait() {
+    std::unique_lock<std::mutex> lock(barrierMutex);
+    const unsigned int gen = curGeneration;
+
+    curCount--;
+
+    if (curCount == 0) {
+      curGeneration++;
+      curCount = barrierCount;
+      barrierCond.notify_all();
+      return;
+    }
+
+    while (gen == curGeneration) {
+      barrierCond.wait(lock);
+    }
+  }
+
+private:
+  std::mutex barrierMutex;
+  std::condition_variable barrierCond;
+  unsigned int curGeneration;
+  unsigned int curCount;
+  const unsigned int barrierCount;
+};
+
 /* Wait for all worker threads */
-void  CmiNodeBarrier(void) {
-  CmiNodeBarrierCount(CmiMyNodeSize(), CMI_NODE_BARRIER);
+void CmiNodeBarrier(void) {
+  static Barrier nodeBarrier(CmiMyNodeSize());
```

Comment:
I'm wondering how static initialization works here when there are multiple threads. AFAIK the static is initialized the first time the function is executed. What happens with multiple threads calling at the same time?

Reply:
Yes, that's right. Since C++11, ["If multiple threads attempt to initialize the same static local variable concurrently, the initialization occurs exactly once."](https://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables)

Reply:
Cool